src/charm.py

```diff
@@ -0,0 +1,249 @@
+        """Get keystone data if ready, otherwise empty dict."""
+        # The double loop is because credentials are in the unit data,
+        # and not all units are guaranteed to have this data.
+        # So we pick the first one that has the data we need.
+        for rel in self.model.relations.get("credentials", []):
+            for unit in rel.units:
+                data = rel.data[unit]
+                if self._is_keystone_data_ready(data):
+                    return data
+        return {}
+
+    def get_resource(self) -> Optional[str]:
+        """Return the path-to-resource or None if the resource is empty.
+
+        Fetch the charm's resource and check if the resource is an empty file
+        or not. If it's empty, return None. Otherwise, return the path to the
+        resource.
+        """
+        try:
+            snap_path = self.model.resources.fetch(RESOURCE_NAME).absolute()
+        except ModelError:
+            logger.debug("cannot fetch charm resource")
+            return None
+
+        if not os.path.getsize(snap_path) > 0:
+            logger.debug("resource is an empty file")
+            return None
+
+        return snap_path
+
+    def install(self) -> None:
+        """Install the necessary resources for the charm."""
+        try:
+            snap_install_or_refresh(self.get_resource(), self.model.config["snap_channel"])
+        except SnapError:
+            self.model.unit.status = BlockedStatus(
+                "Failed to remove/install openstack-exporter snap"
+            )
+
+    def _configure(self, _: ops.HookEvent) -> None:
+        """Configure the charm.
+
+        An idempotent method called as the result of several config or relation changed hooks.
+        It will install the exporter if not already installed or refresh the channel if the it was
+        changed.
+        """
+        self.install()
+
+        upstream_snap = get_installed_snap_service(UPSTREAM_SNAP)
+        if upstream_snap.present:
+            return
+
+        snap_service = get_installed_snap_service(SNAP_NAME)
+        # if cos is not related then we should block and not run anything
+        if not self.model.relations.get("cos-agent"):
+            snap_service.stop()
+            return
+
+        snap_service.configure(
+            {
+                "cloud": CLOUD_NAME,
+                "os-client-config": str(OS_CLIENT_CONFIG),
+                "web": {"listen-address": f":{self.model.config['port']}"},
+                "cache": self.model.config["cache"],
+                "cache-ttl": self.model.config["cache_ttl"],
+            }
+        )
+
+        data = self._get_keystone_data()
+        if not data:
+            logger.info("Keystone credentials are not available, stopping services.")
+            snap_service.stop()
+            return
+
+        logger.info("Keystone credentials are available, starting services.")
+        self._write_cloud_config(data)
+        snap_service.restart_and_enable()
```

Comment:
Not part of the formal review

The flow here was hard for me to understand not being familiar with the context. It wasn't clear what the upstream_snap is, I'm guessing it is a legacy snap that doesn't need to be managed the same way as it is here? Perhaps that could be made clearer

There are also 2 checks here which conditionally stop the snap service. Perhaps group them and don't configure the snap service in those cases

Also, perhaps rename `install` to somehow make it clear that it attempts to remove the upstream snap?

And perhaps then group the configure, writing the clouds yaml and then restart end enable together as well?

Reply:
Another option is to make an explicit function that attempts to remove the upstream snap

Reply:
I do agree with that it's too complex here. Maybe a refactor to decoupling the logic and more explanation can help on that. How do you think @gabrielcocenza ?

Reply:
I don't have a strong opinion on the `install` part. I would be OK leaving as it is because this process of removing the upstream snap is a temporary workaround that I think we should remove it in the future. We might have some clouds before rev28 that needs to have a smooth upgrade, but once that is done (we would need a confirmation from Nobuto), we can remove it.
Maybe raise an exception in the install if the upstream is detected would make things a little bit clear?

My suggestion to refactor is:
- `_write_cloud_config` should be responsible to try to get the openstack credentials and raise an error if keystone credentials are not available.

- I don't see a reason to stop the the exporter if cos-agent is not related, but I'm in favor to keep into block state.

Maybe something like this would reduce the complexity

```python

    def _configure(self, _: ops.HookEvent) -> None:
        """Configure the charm.

        An idempotent method called as the result of several config or relation changed hooks.
        It will install the exporter if not already installed or refresh the channel if the it was
        changed.
        """
        try:
            self.install()
        except UpstreamInstalled:
            return

        snap_service = get_installed_snap_service(SNAP_NAME)

        try:
            self._write_cloud_config()
        except OpenStackCredentialsException:
            snap_service.stop()
            return

        snap_service.configure(
            {
                "cloud": CLOUD_NAME,
                "os-client-config": str(OS_CLIENT_CONFIG),
                "web": {"listen-address": f":{self.model.config['port']}"},
                "cache": self.model.config["cache"],
                "cache-ttl": self.model.config["cache_ttl"],
            }
        )
        snap_service.restart_and_enable()
```

Reply:
that looks better 👍

Reply:
Follow this suggestion in https://github.com/canonical/openstack-exporter-operator/issues/105